### PR TITLE
Tosca parser integration

### PR DESCRIPTION
**Pull request by alexmaslenn**

### CODE REVIEW

File: src/opera_tosca_parser/template/node.py [hunk 1]
```diff
@@ -61,3 +63,16 @@ def resolve_requirements(self, topology: Topology):
             if requirement_occurrences[r.name] > max_occurrences:
                 raise DataError(f"Too many occurrences found for requirement '{r.name}'. Maximum is: {max_occurrences}")
             r.resolve(topology)
+
+    def is_a(self, typ):
```
Comment (anzoman) [inline src/opera_tosca_parser/template/node.py]: Currently all the xOpera TOSCA parser's code is documented and typed I'd like to keep it that way.

Please insert documentation string stub with the similar style as other functions + also add typing everywhere.

File: src/opera_tosca_parser/template/node.py [hunk 2]
```diff
@@ -61,3 +63,16 @@ def resolve_requirements(self, topology: Topology):
...
+        return typ in self.types
+
+    @property
+    def instance(self):
```
Comment (anzoman) [inline src/opera_tosca_parser/template/node.py]: Please insert documentation string stub with the similar style as other functions.

File: src/opera_tosca_parser/template/node.py [hunk 3]
```diff
@@ -61,3 +63,16 @@ def resolve_requirements(self, topology: Topology):
...
+        if not self._instance:
+            raise DataError(f"Node template {self.name} was not instantiated yet")
+        return self._instance
+
+    @instance.setter
+    def instance(self, value):
```
Comment (anzoman) [inline src/opera_tosca_parser/template/node.py]: Please insert documentation string stub with the similar style as other functions.

File: src/opera_tosca_parser/template/relationship.py [hunk 4]
```diff
@@ -27,3 +29,18 @@ def __init__(self, name: str, types: Tuple[str, ...], properties: Dict[str, Valu
 
         # This will be set when the relationship is inserted into a topology
         self.topology: Optional[Topology] = None
+        # This will be set at instantiation time.
+        self._instance = None  # type: ignore
+
+    def is_a(self, typ):
```
Comment (anzoman) [inline src/opera_tosca_parser/template/relationship.py]: Docs missing.

File: src/opera_tosca_parser/template/relationship.py [hunk 5]
```diff
@@ -27,3 +29,18 @@ def __init__(self, name: str, types: Tuple[str, ...], properties: Dict[str, Valu
...
+        return typ in self.types
+
+    @property
+    def instance(self):
```
Comment (anzoman) [inline src/opera_tosca_parser/template/relationship.py]: Docs missing.

File: src/opera_tosca_parser/template/relationship.py [hunk 6]
```diff
@@ -27,3 +29,18 @@ def __init__(self, name: str, types: Tuple[str, ...], properties: Dict[str, Valu
...
+        if not self._instance:
+            raise DataError(f"Relationship template {self.name} was not instantiated yet")
+        return self._instance
+
+    @instance.setter
+    def instance(self, value):
```
Comment (anzoman) [inline src/opera_tosca_parser/template/relationship.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 7]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
     def __str__(self) -> str:
         """Overridden string representation"""
         return f"Value({self.present})[{self._data}]"
+
+    def dump(self):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 8]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        return dict(is_set=self.present, data=self._data)
+
+    def load(self, data):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 9]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        self.present = data["is_set"]
+        self._data = data["data"]
+
+    @property
+    def data(self):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 10]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        if not self.present:
+            raise AssertionError("Accessing an unset value. Bug-fixing ahead ;)")
+        return self._data
+
+    def copy(self):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 11]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        return type(self)(self.type, self.present, copy.deepcopy(self._data))
+
+    def unset(self):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py:75]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 12]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        self._data = None
+        self.present = False
+
+    def eval(self, instance, key):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 13]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        if isinstance(self.data, list):
+            result_list = []
+            for value in self.data:
+                result_list.append(Value.check_eval_function(value, instance))
+            return result_list
+
+        return self.data
+
+    @property
+    def is_function(self):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 14]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        return Value.check_function(self.data)
+
+    @staticmethod
+    def check_function(data):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 15]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        return isinstance(data, dict) and len(data) == 1 and tuple(data.keys())[0] in Value.FUNCTIONS
+
+    @staticmethod
+    def eval_function(data, instance):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

File: src/opera_tosca_parser/value.py [hunk 16]
```diff
@@ -34,3 +38,64 @@ def set(self, data: Any):
...
+        (function_name, params), = data.items()
+        return getattr(instance, function_name)(params)
+
+    @staticmethod
+    def check_eval_function(data, instance):
```
Comment (anzoman) [inline src/opera_tosca_parser/value.py]: Docs missing.

**Review by anzoman — CHANGES_REQUESTED** [review]
Looks okay, @alexmaslenn - just add the docs + typing and we're done here.

**Comment by alexmaslenn** [comment]
Thanks @anzoman for your review. Docs and typing added.

**Review by anzoman — APPROVED** [review]
Thanks @alexmaslenn, let's merge this!

**Pull request merged**